### PR TITLE
[#394] sample mut uninit

### LINKED ISSUE #394: `SampleMut<MaybeUninit<T>>` offers `send` method
## Required Information

This is an API bug.

For the user it is possible to send an uninitialized sample, this must be prevented by the API otherwise it will induce undefined behavior in another application.

## Example

```rust
// bad
let sample = publisher.loan_uninit()?;
sample.send()?; // must be a compile failure

// good
let mut sample = publisher.loan_uninit()?;
sample.write_payload(123);
let sample = unsafe { sample.assume_init() };
sample.send()?;
```

### PR DESCRIPTION
**Pull request by elfenpiff**
## Notes for Reviewer

 * Fixes the problem that uninitialized samples can be sent by introducing a `SampleMutUninit` struct that does not offer a `send` method
 * Renames the method `send_sample` into `send` since it was weird to read `send_sample(std::move(sample))`, this is much cleaner `send(std::move(sample))`.
 * The introduction of the new class needed an adjustment on the C API level - here we always use uninitialized samples
    * `SampleMut` has no C binding, the C++ implementation is done directly via the C API, otherwise it would be much more code on the FFI layer

## Pre-Review Checklist for the PR Author

1. [x] Add sensible notes for the reviewer
1. [x] PR title is short, expressive and meaningful
1. [x] Relevant issues are linked in the [References](#references) section
1. [x] Every source code file has a copyright header with `SPDX-License-Identifier: Apache-2.0 OR MIT`
1. [x] Branch follows the naming format (`iox2-123-introduce-posix-ipc-example`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`[#123] Add posix ipc example`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Assign PR to reviewer
1. [x] All checks have passed (except `task-list-completed`)

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx2/blob/main/doc/release-notes/iceoryx2-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Unit tests have been written for new behavior
- [x] Public API is documented
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## References

Closes #394

### CODE REVIEW

File: iceoryx2-bb/log/src/lib.rs [hunk 1]
```diff
@@ -161,7 +161,7 @@ static DEFAULT_LOGGER: logger::log::Logger = logger::log::Logger::new();
 #[cfg(not(any(feature = "logger_log", feature = "logger_tracing")))]
 static DEFAULT_LOGGER: logger::console::Logger = logger::console::Logger::new();
 
-const DEFAULT_LOG_LEVEL: u8 = LogLevel::Trace as u8;
+const DEFAULT_LOG_LEVEL: u8 = LogLevel::Warn as u8;
```
Comment (orecham) [inline iceoryx2-bb/log/src/lib.rs:164]: I guess this is based on #393

**Review by orecham — COMMENTED** [review]
@elfenpiff Looks good. Ping me for approval after rebasing on `main` or merging #393.

File: iceoryx2-bb/log/src/lib.rs (hunk 1 above)
Comment (elfenpiff) [reply iceoryx2-bb/log/src/lib.rs:164]: Yes it is ... I forget to mention this.

**Review by orecham — APPROVED** [review]

File: doc/release-notes/iceoryx2-unreleased.md [hunk 2]
```diff
@@ -458,3 +461,22 @@
...
+20. `SampleMutUninit` without `send` as replacement for `SampleMut<MaybeUninit<Payload>>`
+
+    ```rust
+    // old
+    let sample: SampleMut<zero_copy::Service, MaybeUninit<u64>> = publisher.loan_uninit()?;
+    let sample = sample.write_payload(123);
+    let sample.send()?;
+
+    // new
+    let sample: SampleMutUninit<zero_copy::Service, MaybeUninit<u64>> = publisher.loan_uninit()?;
```
Comment (elBoberido) [inline doc/release-notes/iceoryx2-unreleased.md:474]: Shouldn't this be just
```suggestion
    let sample: SampleMutUninit<zero_copy::Service, u64> = publisher.loan_uninit()?;
```

File: iceoryx2-ffi/ffi/src/api/sample_mut_uninit.rs [hunk 3]
```diff
@@ -27,21 +27,21 @@ use super::UninitPayloadFfi;
 
 // BEGIN types definition
 
-pub(super) union SampleMutUnion {
-    ipc: ManuallyDrop<SampleMut<ipc::Service, UninitPayloadFfi, UserHeaderFfi>>,
-    local: ManuallyDrop<SampleMut<local::Service, UninitPayloadFfi, UserHeaderFfi>>,
+pub(super) union SampleMutUninitUnion {
+    ipc: ManuallyDrop<SampleMutUninit<ipc::Service, UninitPayloadFfi, UserHeaderFfi>>,
+    local: ManuallyDrop<SampleMutUninit<local::Service, UninitPayloadFfi, UserHeaderFfi>>,
```
Comment (elBoberido) [inline iceoryx2-ffi/ffi/src/api/sample_mut_uninit.rs:32]: If it's an either or on the C API, I think the going with just `SampleMut` makes more sense, else one has to call `iox2_sample_mut_uninit_send`, which is definitely the wrong message we want to spread.

**Review by elBoberido — DISMISSED** [review]
The request changes is just to make you aware of my comments. If you dismiss them, I'm fine with merging.

File: doc/release-notes/iceoryx2-unreleased.md (hunk 2 above)
Comment (elfenpiff) [reply doc/release-notes/iceoryx2-unreleased.md:474]: Nope, `MaybeUninit` is still required. The reason is that `Payload: ?Sized` but you cannot put a type into a maybe uninit where the size is not known at compile time.
Internally, the `SampleMutUninit` just holds a `SampleMut<..., Payload>`. If I pack this there into a `MaybeUninit` it fails to compile for slices.

Comment (elBoberido) [reply doc/release-notes/iceoryx2-unreleased.md:474]: I'll have a look at this later today. If you want to merge this sooner, we can also do it and I create a new PR if I find a solution

**Review by elBoberido — APPROVED** [review]

**Pull request merged**